boofuzz/base_socket_connection.py

```diff
@@ -0,0 +1,59 @@
+from __future__ import absolute_import
+
+import abc
+import math
+import os
+import socket
+import struct
+
+
+from future.utils import with_metaclass
+from . import itarget_connection
+
+
+def _seconds_to_sockopt_format(seconds):
+    """Convert floating point seconds value to second/useconds struct used by UNIX socket library.
+    For Windows, convert to whole milliseconds.
+    """
+    if os.name == "nt":
+        return int(seconds * 1000)
+    else:
+        microseconds_per_second = 1000000
+        whole_seconds = int(math.floor(seconds))
+        whole_microseconds = int(math.floor((seconds % 1) * microseconds_per_second))
+        return struct.pack("ll", whole_seconds, whole_microseconds)
+
+
+class BaseSocketConnection(with_metaclass(abc.ABCMeta, itarget_connection.ITargetConnection)):
+    """This class serves as a base for a number of Connections over sockets.
+
+    Args:
+        send_timeout (float): Seconds to wait for send before timing out. Default 5.0.
+        recv_timeout (float): Seconds to wait for recv before timing out. Default 5.0.
+    """
+
+    def __init__(self, socket, send_timeout, recv_timeout):
+        self._send_timeout = send_timeout
+        self._recv_timeout = recv_timeout
+
+        self._sock = None
+
+    def close(self):
+        """
+        Close connection to the target.
+
+        Returns:
+            None
+        """
+        self._sock.close()
+
+    @abc.abstractmethod
```

Comment:
Shouldn't all these members just tickle through from the baseclass? At least in my ipython shell, this works as expected:

    In [1]: from boofuzz import BaseSocketConnection                                                                    
    In [2]: class t(BaseSocketConnection): 
 
      ...:     def __init__(self): 
      ...:         super().__init__(5.0, 5.0) 
      ...:                                                                                                             
    In [3]: s = t()                                                                                                     
    ---------------------------------------------------------------------------
    TypeError                                 Traceback (most recent call last)
    <ipython-input-3-e764f24fca13> in <module>
    ----> 1 s = t()
    TypeError: Can't instantiate abstract class t with abstract methods info, open, recv, send
    In [4]:  

As for recv(), personally, I'd rather see this implemented in the respective classes and not in the base class as this would prompt developers to write some docs on why recv() is unimplemented thus making leaving it unimplemented a conscious decision.

Reply:
Sorry my bad, I completely forgot about `ITargetConnection`.
You are right, keep it as it is.